src/library_wasm_worker.js

```diff
@@ -0,0 +1,391 @@
+    + '_wasm_workers[0] = this;\n'
+    + 'addEventListener("message", __wasm_worker_appendToQueue);\n'
+    + '}\n',
+  _emscripten_create_wasm_worker_with_tls: function(stackLowestAddress, stackSize, tlsAddress) {
+#if ASSERTIONS
+    assert(stackLowestAddress % 16 == 0);
+    assert(stackSize % 16 == 0);
+#endif
+    let worker = _wasm_workers[_wasm_workers_id] = new Worker(
+#if WASM_WORKERS == 2 // WASM_WORKERS=2 mode embeds .ww.js file contents into the main .js file as a Blob URL. (convenient, but not CSP security safe, since this is eval-like)
+      __wasmWorkerBlobUrl
+#elif MINIMAL_RUNTIME // MINIMAL_RUNTIME has a structure where the .ww.js file is loaded from the main HTML file in parallel to all other files for best performance
+      Module['$wb'] // $wb="Wasm worker Blob", abbreviated since not DCEable
+#else // default runtime loads the .ww.js file on demand.
+      locateFile('{{{ WASM_WORKER_FILE }}}')
+#endif
+    );
+    // Craft the Module object for the Wasm Worker scope:
+    worker.postMessage({
+      '$ww': _wasm_workers_id, // Signal with a non-zero value that this Worker will be a Wasm Worker, and not the main browser thread.
+#if MINIMAL_RUNTIME
+      'wasm': Module['wasm'],
+      'js': Module['js'],
+      'mem': wasmMemory,
+#else
+      'wasm': wasmModule,
+      'js': Module['mainScriptUrlOrBlob'] || _scriptDir,
+      'wasmMemory': wasmMemory,
+#endif
+      'sb': stackLowestAddress, // sb = stack base
+      'sz': stackSize,          // sz = stack size
+#if !WASM_WORKERS_NO_TLS
+      'tb': tlsAddress,         // tb = TLS base
+#endif
+    });
+    worker.addEventListener('message', __wasm_worker_runPostMessage);
+    return _wasm_workers_id++;
+  },
+  _emscripten_create_wasm_worker_no_tls__deps: ['_emscripten_create_wasm_worker_with_tls'],
+  _emscripten_create_wasm_worker_no_tls: function(stackLowestAddress, stackSize) {
+    return __emscripten_create_wasm_worker_with_tls(stackLowestAddress, stackSize, 0, 0);
+  },
+
+  emscripten_terminate_wasm_worker: function(id) {
+#if ASSERTIONS
+    assert(id != 0, 'emscripten_terminate_wasm_worker() cannot be called with id=0!');
+#endif
+    if (_wasm_workers[id]) {
+      _wasm_workers[id].terminate();
+      delete _wasm_workers[id];
+    }
+  },
+
+  emscripten_terminate_all_wasm_workers: function() {
+#if ASSERTIONS
+    assert(!ENVIRONMENT_IS_WASM_WORKER, 'emscripten_terminate_all_wasm_workers() cannot be called from a Wasm Worker: only the main browser thread has visibility to terminate all Workers!');
+#endif
+    Object.values(_wasm_workers).forEach((worker) => {
+      worker.terminate();
+    });
+    _wasm_workers = {};
+  },
+
+  emscripten_current_thread_is_wasm_worker: function() {
+#if WASM_WORKERS
+    return ENVIRONMENT_IS_WASM_WORKER;
+#else
+    // implicit return 0;
+#endif
+  },
+
+  emscripten_wasm_worker_self_id: function() {
+    return Module['$ww'];
+  },
+
+  emscripten_wasm_worker_post_function_v__deps: ['$dynCall'],
```

Comment:
Why do we need a dependency on `$dynCall` if it is not called in the function body?

Would it be feasible to expose functionality only for posting `void (*)(void*)` function pointers, then layer the dynamic typing and dispatch of other function signatures on top of that a a separate library?

Reply:
> Why do we need a dependency on $dynCall if it is not called in the function body?
Good observation.

These were posted here for the receiving side's behalf, `_wasm_worker_runPostMessage` would do a dynCall. But now the `getWasmTableEntry` functionality has landed in between, so `_wasm_worker_runPostMessage` has migrated to use that, so declaring these kinds of deps are no longer needed. Removed.

Reply:
> Would it be feasible to expose functionality only for posting void (*)(void*) function pointers, then layer the dynamic typing and dispatch of other function signatures on top of that a a separate library?

That would be awkward for users, since it would require dynamic memory allocation from the caller to receiver thread. There is `emscripten_wasm_worker_post_function_sig()` that can be used to post arbitrary function signatures (without dynamic memory allocation)

Reply:
This wouldn't always require dynamic allocation. For synchronous proxying a stack allocation would be sufficient, for example. And if we extracted the dynamic typing and dispatch code currently in threading.h out into a standalone library, it could easily be used here and anywhere else we need to proxy functions.

Reply:
> For synchronous proxying

The `emscripten_wasm_worker_post_function*` function family only does a `postMessage()`, so it cannot do synchronous proxying.

> it could easily be used here and anywhere else we need to proxy functions.

The vocabulary I wanted to build here is that this function set exposes the JS native `postMessage()` capability to post messages to the receiver's JS event queue.

The "proxying" vocabulary is generally about routing a function call to be executed on another thread's context via the wasm heap, and the time when the receiver shall execute the function differs: with proxying, the receiver executes the messages at pthread cancellation points. With postMessage(), the receive will execute the function only after yielding to the JS event loop.